### ISSUE TITLE
How to get Neg samples?

### ISSUE THREAD

**Issue by Valeria281856** [fawnliu/TRIS#10]
I have read the code L199-L219 from ReferDataset.py, but I donnot understand : Are there different objects of different referring sentences  for the same image. I think there are same objects for the same image, even thouth we have different referring sentences to represent the same object.

**Comment by fawnliu** [comment]
Hi, @xiao-11 
We implement this part of code in [L190-L229 of dataset/ReferDataset.py](https://github.com/fawnliu/TRIS/blob/c6666b3b89a31504a23f39710e904e1335e23e91/dataset/ReferDataset.py#L190). 


First, we need to find all ref_ids that belong to the current images. Then, we store the ref_ids that are different from the current ref_id (`this_ref_id`) in `neg_index`. After that, we can sample some negative samples based on the ref_ids in `neg_index`.


```python
if self.negative_samples > 0:
      ###########
      img2ref = self.refer.imgToRefs[this_img_id[0]]
      neg_index = []
      for item in img2ref:
          t_ref_id = item['ref_id']
          t_category_id = item['category_id']
          try:
              if t_ref_id != this_ref_id:  # and this_category_id == t_category_id
                  neg_index.append(self.refid2index[t_ref_id])
          except: ### for refcocog google, its refindex is not match
              break 
              import pdb
              pdb.set_trace() 
      ###########

      if len(neg_index) > 0:
          neg_sents = []
          neg_word_ids = []
          ## random select negtive samples from same random index 
          # n_index = neg_index[np.random.choice(len(neg_index))]
          while len(neg_sents) < self.negative_samples:
              ## different random index 
              n_index = neg_index[np.random.choice(len(neg_index))]
              choice_sent = np.random.choice(len(self.input_ids[n_index]))
              neg_word_ids.append(self.input_ids[n_index][choice_sent])
              neg_sents.append(self.all_sentences[n_index][choice_sent])
          neg_word_ids = torch.cat(neg_word_ids, dim=0)
      else:
          # random index, then randomly select one sentence 
          neg_sents = []
          neg_word_ids = []
          while len(neg_sents) < self.negative_samples:
              n_index = np.random.choice(len(self.input_ids))
              choice_sent = np.random.choice(len(self.input_ids[n_index]))
              tmp_sent = self.all_sentences[n_index][choice_sent]
              if tmp_sent != sentences:
                  neg_sents.append(tmp_sent)
                  neg_word_ids.append(self.input_ids[n_index][choice_sent])
          neg_word_ids = torch.cat(neg_word_ids, dim=0)
